Fixed the Unit Tests on Windows

jmittert opened the pull request:
This fixes all the unittests on Windows!

```
S:\llbuild> C:\Python27\python.exe C:/Library/Developer/Toolchains/unknown-Asserts-development.xctoolchain/usr/bin/llvm-lit.py --param build_mode=Debug S:/build/Ninja-DebugAssert/llbuild-windows-amd64/tests --filter "unit" -sv                                                      
-- Testing: 64 of 187 tests, 12 threads --                                                                                                  
Testing: 0 .. 10.. 20.. 30.. 40.. 50.. 60.. 70.. 80.. 90..                                                                                  
Testing Time: 22.26s                                                                                                                        
  Expected Passes    : 64  
```

jmittert commented on lib/Basic/PlatformUtility.cpp:
```diff
@@ -42,11 +44,68 @@ int sys::close(int fileHandle) {
-  // We deliberately ignore lstat on Windows, and delegate
-  // to stat.
-  return ::_stat(fileName, buf);
+  llvm::SmallVector<UTF16, 20> wfilename;
```
This is definitely not a _good_ implementation of lstat, but Windows doesn't provide it and it's good enough for the purposes of checking if file is a directory/regular file/symlink so llbuild knows to recurse into it or not.

jmittert commented on unittests/Basic/FileSystemTest.cpp:
```diff
@@ -40,7 +40,7 @@ TEST(FileSystemTest, basic) {
     llvm::raw_fd_ostream os(tempPath.str(), ec, llvm::sys::fs::F_Text);
     EXPECT_FALSE(ec);
-    os << "Hello, world!\n";
+    os << "Hello, world!";
```
This was causing problems with `\r\n`. Rather than `ifdef`ing, I figured it should be fine to just not check the newlines

jmittert commented on unittests/Basic/FileSystemTest.cpp:
```diff
@@ -54,15 +54,15 @@ TEST(FileSystemTest, basic) {
 
 TEST(FileSystemTest, testRecursiveRemoval) {
-  TmpDir rootTempDir{ __FUNCTION__ };
+  TmpDir rootTempDir(__func__);
```
`__FUNCTION__` in MSVC doesn't give the same thing as `__FUNCTION__` in Clang. `__func__` however is standardized in C++11 and is the same thing as `__FUNCTION__` with Clang

jmittert commented on unittests/BuildSystem/BuildSystemFrontendTest.cpp:
```diff
@@ -343,6 +345,20 @@ TEST_F(BuildSystemFrontendTest, commandSkippingFailure) {
+shouldCommandStart: 2
+commandStarted: 2
+commandProcessStarted: 2
+commandProcessFinished: 2: 1
```
Should be the same except that Windows gives a different error code

jmittert commented on unittests/BuildSystem/BuildSystemFrontendTest.cpp:
```diff
@@ -441,11 +457,16 @@ TEST_F(BuildSystemFrontendTest, commandSkipping_NonExternalCommands_symlink) {
         contents: "x"
     2:
-        tool: symlink
+        tool: shell
```
Windows seems to be unable to rm a symlink if the target doesn't exist though it doesn't have an issue creating one.

jmittert replied:
Can you expand on what you mean by "handle that internally to the shell tool"?

jakepetroules replied:
I know it's with the deletion, but isn't there a Win32 API we can call to delete a symlink without attempting to follow it?

jmittert replied:
I don't follow... there are Win32 APIs to do exactly that, but gnu msys rm doesn't do it properly. It says it's using lstat, but whatever implementation of lstat it's using doesn't seem to be correct.  Unless you're alluding to providing a RemoveCommand that does this properly?

jakepetroules replied:
Ohhh, sorry, I completely misread the diff due to the way git is presenting it here. Disregard all this :)

jmittert commented:
Alright, remove the extra formatting changes, changed all winapi calls to the W ones and addressed the remaining comments. I don't have access to a macOS machine right now, but the tests look good on the Vagrant box so... *crosses fingers*.

dmbryson commented:
@swift-ci please smoke test

jakepetroules commented:
> Alright, removed the extra formatting changes, changed all winapi calls to the W ones and addressed the remaining comments. I don't have access to a macOS machine right now, but the tests look good on the Vagrant box so... _crosses fingers_.

Our CI will validate the macOS build, no worries.

jakepetroules commented:
@swift-ci please smoke test
